wcfsetup/install/files/lib/system/package/plugin/AbstractFileDeletePackageInstallationPlugin.class.php

```diff
@@ -0,0 +1,306 @@
+
+    /**
+     * Returns the name of the column in the log table returned by `getLogTableName()` that contains
+     * the names of the relevant files.
+     */
+    abstract protected function getFilenameTableColumn(): string;
+
+    protected function getPipName(): string
+    {
+        return $this->getXsdFilename();
+    }
+
+    /**
+     * Returns the actual absolute path of the given file.
+     */
+    protected function getFilePath(string $filename, string $application): string
+    {
+        return Application::getDirectory($application) . $filename;
+    }
+
+    /**
+     * @inheritDoc
+     */
+    protected function handleDelete(array $items)
+    {
+        $sql = "SELECT  packageID
+                FROM    {$this->getLogTableName()}
+                WHERE   {$this->getFilenameTableColumn()} = ?
+                    AND application = ?
+                    AND packageID = ?";
+        $searchStatement = WCF::getDB()->prepare($sql);
+
+        $sql = "DELETE FROM {$this->getLogTableName()}
+                WHERE       packageID = ?
+                        AND {$this->getFilenameTableColumn()} = ?";
+        $deleteStatement = WCF::getDB()->prepare($sql);
+
+        foreach ($items as $item) {
+            $file = $item['value'];
+            $application = 'wcf';
+            if (!empty($item['attributes']['application'])) {
+                $application = $item['attributes']['application'];
+            } elseif ($this->installation->getPackage()->isApplication) {
+                $application = Package::getAbbreviation($this->installation->getPackage()->package);
+            }
+
+            $searchStatement->execute([
+                $file,
+                $application,
+                $this->installation->getPackageID(),
+            ]);
+
+            $filePackageID = $searchStatement->fetchSingleColumn();
+            if ($filePackageID !== false && $filePackageID != $this->installation->getPackageID()) {
+                throw new \UnexpectedValueException(
+                    "'{$file}' does not belong to package '{$this->installation->getPackage()->package}'
+                    but to package '" . PackageCache::getInstance()->getPackage($filePackageID)->package . "'."
+                );
+            }
+
+            $filePath = $this->getFilePath($file, $application);
+            if (\file_exists($filePath)) {
+                \unlink($filePath);
+            }
+
+            $deleteStatement->execute([
+                $this->installation->getPackageID(),
+                $file,
+            ]);
+        }
+    }
+
+    /**
+     * @inheritDoc
+     */
+    final protected function import(array $row, array $data)
+    {
+        // Does nothing, imports are not supported.
+    }
+
+    /**
+     * @inheritDoc
+     */
+    protected function prepareImport(array $data)
+    {
+        return $data;
+    }
+
+    /**
+     * @inheritDoc
+     */
+    protected function findExistingItem(array $data)
```

Comment:
Maybe also this one, because it's related to `import()`? Might apply to others as well.